crates/lsp/src/workspaces.rs

```diff
@@ -0,0 +1,174 @@
+        match Self::url_to_path(url)? {
+            Some(path) => {
+                tracing::trace!("Removing workspace settings: {}", path.display());
+                Ok(self.path_to_settings_resolver.remove(&path))
+            }
+            None => {
+                tracing::warn!("Ignoring non-file workspace URL: {url}");
+                Ok(None)
+            }
+        }
+    }
+
+    /// Return the appropriate [`Settings`] for a given document [`Url`].
+    pub(crate) fn settings_for_url(&self, url: &Url) -> &Settings {
+        if let Ok(Some(path)) = Self::url_to_path(url) {
+            return self.settings_for_path(&path);
+        }
+
+        // For `untitled` schemes, we have special behavior.
+        // If there is exactly 1 workspace, we resolve using a path of
+        // `{workspace_path}/untitled` to provide relevant settings for this workspace.
+        if url.scheme() == "untitled" && self.path_to_settings_resolver.len() == 1 {
+            tracing::trace!("Using workspace settings for 'untitled' URL: {url}");
+            let workspace_path = self.path_to_settings_resolver.keys().next().unwrap();
+            let path = workspace_path.join("untitled");
+            return self.settings_for_path(&path);
+        }
+
+        tracing::trace!("Using default settings for non-file URL: {url}");
+        self.path_to_settings_resolver.fallback().fallback()
+    }
+
+    /// Reloads all workspaces matched by the [`Url`]
+    ///
+    /// This is utilized by the watched files handler to reload the settings
+    /// resolver whenever an `air.toml` is modified.
+    pub(crate) fn reload_workspaces_matched_by_url(&mut self, url: &Url) {
+        let path = match Self::url_to_path(url) {
+            Ok(Some(path)) => path,
+            Ok(None) => {
+                tracing::trace!("Ignoring non-`file` changed URL: {url}");
+                return;
+            }
+            Err(error) => {
+                tracing::error!("Failed to reload workspaces associated with {url}:\n{error}");
+                return;
+            }
+        };
+
+        if !path.ends_with("air.toml") {
+            // We could get called with a changed file that isn't an `air.toml` if we are
+            // watching more than `air.toml` files
+            tracing::trace!("Ignoring non-`air.toml` changed URL: {url}");
+            return;
+        }
+
+        for (workspace_path, settings_resolver) in self.path_to_settings_resolver.matches_mut(&path)
+        {
+            tracing::trace!("Reloading workspace settings: {}", workspace_path.display());
+
+            settings_resolver.clear();
+
+            if let Err(error) = settings_resolver.load_from_paths(&[workspace_path]) {
+                tracing::error!(
+                    "Failed to reload workspace settings for {path}:\n{error}",
+                    path = workspace_path.display(),
+                    error = error
+                );
+            }
+        }
```

Comment:
I find this "refresh" behavior kind of awesome. We mutably iterate over the workspace specific resolvers that are associated with the updated `air.toml` and _update them in place_ by clearing them and then rediscovering the `air.toml`s within them